client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -0,0 +1,238 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use libp2p::core::multiaddr::{Multiaddr, Protocol};
+use rand::seq::SliceRandom;
+use std::{
+	clone::Clone,
+	collections::HashMap,
+	convert::AsRef,
+};
+
+use sp_authority_discovery::AuthorityId;
+use sc_network::PeerId;
+
+/// The maximum number of authority connections initialized through the authority discovery module.
+///
+/// In other words the maximum size of the `authority` peerset priority group.
+const MAX_NUM_AUTHORITY_CONN: usize = 10;
+
+/// Cache for [`AuthorityId`] -> [`Vec<Multiaddr>`] and [`PeerId`] -> [`AuthorityId`] mappings.
+pub(super) struct AddrCache {
+	authority_id_to_addresses: HashMap<AuthorityId, Vec<Multiaddr>>,
+	peer_id_to_authority_id: HashMap<PeerId, AuthorityId>,
+}
+
+impl AddrCache {
+	pub fn new() -> Self {
+		AddrCache {
+			authority_id_to_addresses: HashMap::new(),
+			peer_id_to_authority_id: HashMap::new(),
+		}
+	}
+
+	/// Inserts the given [`AuthorityId`] and [`Vec<Multiaddr>`] pair for future lookups by
+	/// [`AuthorityId`] or [`PeerId`].
+	pub fn insert(&mut self, authority_id: AuthorityId, mut addresses: Vec<Multiaddr>) {
+		if addresses.is_empty() {
+			return;
+		}
+
+		// Insert into `self.peer_id_to_authority_id`.
+		let peer_ids = addresses.iter()
+			.map(|a| peer_id_from_multiaddr(a))
+			.filter_map(|peer_id| peer_id);
+		for peer_id in  peer_ids {
+			self.peer_id_to_authority_id.insert(peer_id, authority_id.clone());
+		}
+
+		// Insert into `self.authority_id_to_addresses`.
+		addresses.sort_unstable_by(|a, b| a.as_ref().cmp(b.as_ref()));
+		self.authority_id_to_addresses.insert(authority_id, addresses);
+	}
+
+	/// Returns the number of authority IDs in the cache.
+	pub fn num_ids(&self) -> usize {
+		self.authority_id_to_addresses.len()
+	}
+
+	/// Returns the addresses for the given [`AuthorityId`].
+	pub fn get_addresses_by_authority_id(&self, authority_id: &AuthorityId) -> Option<&Vec<Multiaddr>> {
+		self.authority_id_to_addresses.get(&authority_id)
+	}
+
+	/// Returns the [`AuthorityId`] for the given [`PeerId`].
+	pub fn get_authority_id_by_peer_id(&self, peer_id: &PeerId) -> Option<&AuthorityId> {
+		self.peer_id_to_authority_id.get(peer_id)
+	}
+
+	/// Returns a single address for a random subset (maximum of [`MAX_NUM_AUTHORITY_CONN`]) of all
+	/// known authorities.
+	pub fn get_random_subset(&self) -> Vec<Multiaddr> {
+		let mut rng = rand::thread_rng();
+
+		let mut addresses = self
+			.authority_id_to_addresses
+			.iter()
+			.filter_map(|(_authority_id, addresses)| {
+				debug_assert!(!addresses.is_empty());
+				addresses
+					.choose(&mut rng)
+			})
+			.collect::<Vec<&Multiaddr>>();
+
+		addresses.sort_unstable_by(|a, b| a.as_ref().cmp(b.as_ref()));
+		addresses.dedup();
+
+		addresses
+			.choose_multiple(&mut rng, MAX_NUM_AUTHORITY_CONN)
+			.cloned()
+			.cloned()
```

Comment:
is this to make sure it's *really* cloned? :)

Reply:
You never know :P

Jokes aside: `addresses` is a `Vec<&Multiaddr>`. `choose_multiple` returns [an iterator that returns a reference](https://docs.rs/rand/0.7.3/rand/seq/struct.SliceChooseIter.html#impl-Iterator) thus returning a `&&Multiaddr`. Instead of the `clone().clone()` I use `map(|a| (**a).clone())` now. More than happy for prettier suggestions @ordian!